### ISSUE TITLE
Traversal of impassable code

### ISSUE THREAD

**Issue by Thomas-Shephard** [python/cpython#107314]
# Bug report

Python wrongly traverses impassable code but doesn't execute it.

## Bug shown with breakpoints
The issue can easily be seen when using an IDE like Pycharm and using debug mode. In the following code example, if you place a breakpoint on the line with impassable code - the application will hit the breakpoint and pause on the line with impassable code as shown in the image.

```
if False:
    print("a")
```

![image](https://github.com/python/cpython/assets/14152994/6d4f611f-94f1-4c61-a4a1-626cc8c7ac4b)

## Bug shown with Coveragepy
This issue also causes problems with code coverage tools like Coveragepy because it will report that lines like the one above have been traversed and hence that a test has provided coverage to the impassable code as shown in the image.

```
class TestClass:
    pass

if False:
    print("a")
```

```
from unittest import TestCase


class MainTest(TestCase):
    def test_main(self):
        from main import TestClass
```

`coverage run -m unittest discover`
`coverage html`

![image](https://github.com/python/cpython/assets/14152994/5c53cb95-4cf4-4e9b-a899-3cde444a7f5a)

# Your environment

- CPython versions tested on: 3.9.13
- Operating system and architecture: Windows 64bit

**Comment by gaogaotiantian** [comment]
The direct reason for this issue was because for all modules, Python adds an implicit `return None` at the end of the module, which has the same line number as the last line.

```python
  1           0 LOAD_BUILD_CLASS
              2 LOAD_CONST               0 (<code object TestClass at 0x7f1cf6801030, file "<dis>", line 1>)
              4 LOAD_CONST               1 ('TestClass')
              6 MAKE_FUNCTION            0
              8 LOAD_CONST               1 ('TestClass')
             10 CALL_FUNCTION            2
             12 STORE_NAME               0 (TestClass)

  5          14 LOAD_CONST               2 (None)
             16 RETURN_VALUE

Disassembly of <code object TestClass at 0x7f1cf6801030, file "<dis>", line 1>:
  1           0 LOAD_NAME                0 (__name__)
              2 STORE_NAME               1 (__module__)
              4 LOAD_CONST               0 ('TestClass')
              6 STORE_NAME               2 (__qualname__)

  2           8 LOAD_CONST               1 (None)
             10 RETURN_VALUE
```

So the executed line is the hidden `return None`, not `print('a')`. The inexecutable statement just happened to be at the end of the file.

I'm not sure if/how we should fix this, this is related to the compiler, not the tracing system. If the compiler adds an implicit line at the end of file, I don't see a clean way the tracing system deals with it.

**Comment by gaogaotiantian** [comment]
This issue is probably very similar to https://github.com/python/cpython/issues/103971. You can repro it with function in 3.9 as well. Confirmed that this is fixed in 3.11 and main. As 3.9 does not take bug fixes anymore, we won't do anything about it.

Please let us know if you can repro this on 3.11+.

**Comment by terryjreedy** [comment]
When I run the 2 lines `if 0:\n    print('a')` in 3.12 IDLE with debugger on and breakpoint set on the print, the print does not execute.  Adding a 3rd line `print('b')`, execution stops on the latter or prints and exist depending on the presence of a breakpoint.
Ditto in 3.11.
